### PR TITLE
Add String#toLowerCase and String#toUpperCase

### PR DESCRIPTION
**Pull request by MaxGraey**
- [x] Fully valid case mapping with latest Unicode 12.1.0 support
- [x] Basic tests
- [x] Full unicode range stress tests
- [x] More optimizations

[benchmark](https://github.com/AssemblyScript/assemblyscript/pull/965#issuecomment-557523336)

### CODE REVIEW

**Comment by MaxGraey** [comment]
Simple benchmark: https://webassembly.studio/?f=2wns3rmaksv

results for Chrome 78
```
AS toLowerCase: 61.492919921875ms
AS toUpperCase: 67.710693359375ms
JS toLowerCase: 91.915771484375ms
JS toUpperCase: 88.08203125ms
```

**Comment by dcodeIO** [comment]
On my machine:
```
AS toLowerCase: 43.217041015625ms
AS toUpperCase: 52.4541015625ms
JS toLowerCase: 95.34619140625ms
JS toUpperCase: 93.244140625ms
```

**Comment by MaxGraey** [comment]
Great!

**Comment by MaxGraey** [comment]
I little bit improve bench and reduce `codePointAt` influences. Now more clear results:
https://webassembly.studio/?f=xpio8bmtad

```
AS toLowerCase: 34.282958984375ms
AS toUpperCase: 50.489990234375ms
JS toLowerCase: 86.980712890625ms
JS toUpperCase: 79.905029296875ms
```

**Comment by MaxGraey** [comment]
Updated bench: https://webassembly.studio/?f=olg80l4l7nl

with --runtime none (Chrome 78)
```
AS toLowerCase: 38.075927734375ms
AS toUpperCase: 53.0908203125ms
JS toLowerCase: 82.6279296875ms
JS toUpperCase: 77.438720703125ms
```
with --runtime half (Chrome 78)
```
AS toLowerCase: 99.2978515625ms
AS toUpperCase: 98.48193359375ms
JS toLowerCase: 91.281005859375ms
JS toUpperCase: 77.448974609375ms
```

**Comment by MaxGraey** [comment]
But pretty interesting results on Chrome Canary 80.0.3973.0

```
AS toLowerCase: 38.1259765625ms
AS toUpperCase: 52.248779296875ms
JS toLowerCase: 153.882080078125ms
JS toUpperCase: 147.2509765625ms
```

**Comment by MaxGraey** [comment]
New bench results: https://webassembly.studio/?f=q0n8gcizpnf

`--runtime none` (Chrome 78):
```
AS toLowerCase: 35.214111328125ms
AS toUpperCase: 35.929931640625ms
JS toLowerCase: 82.121826171875ms
JS toUpperCase: 77.5810546875ms
```

`--runtime half` (Chrome 78):
```
AS toLowerCase: 89.85107421875ms
AS toUpperCase: 85.462158203125ms
JS toLowerCase: 85.673828125ms
JS toUpperCase: 78.34423828125ms
```

**Comment by dcodeIO** [comment]
On my machine:

`--runtime none`

```
AS toLowerCase: 25.455078125ms
AS toUpperCase: 29.87109375ms
JS toLowerCase: 100.968017578125ms
JS toUpperCase: 92.3291015625ms
```

`--runtime half`

```
AS toLowerCase: 83.18310546875ms
AS toUpperCase: 83.131103515625ms
JS toLowerCase: 100.512939453125ms
JS toUpperCase: 94.1240234375ms
```

**Comment by mathiasbynens** [comment]
I wonder how https://gist.github.com/mathiasbynens/a37e3f3138069729aa434ea90eea4a3c#file-readme-md compares to this :)

**Comment by MaxGraey** [comment]
@mathiasbynens I compare our case mapping exactly with [`v8` for `0 - 0x10FFFF` range](https://github.com/AssemblyScript/assemblyscript/pull/965/files#diff-6904b50499ee3171148ef0c119c38b27R607)

**Comment by MaxGraey** [comment]
@mathiasbynens Also in this PR I used this [SpecialCasing.txt](https://unicode.org/Public/UNIDATA/SpecialCasing.txt) which relate to latest Unicode `12.1.0`

**Comment by MaxGraey** [comment]
@mathiasbynens Added some special case maps for basic tests from your gist for make sure:
https://github.com/AssemblyScript/assemblyscript/pull/965/commits/f7cd9f7f09c41a60494a2bacec3ea0e3b906cd6f#diff-6904b50499ee3171148ef0c119c38b27

**Comment by mathiasbynens** [comment]
@MaxGraey Could you please file a bug at crbug.com/v8 with these perf findings? The TurboFan team will take a look.

**Comment by MaxGraey** [comment]
@mathiasbynens Oh sure! Will do.
And sorry for this noise

**Comment by mathiasbynens** [comment]
@MaxGraey Don't be sorry; it sounds like you've spotted a bug, so thank you!

File: tests/compiler.js [hunk 1]
```diff
@@ -309,6 +309,14 @@ function testInstantiate(basename, binaryBuffer, name, glue) {
...
           trace: function(msg, n) {
             console.log("  trace: " + getString(msg) + (n ? " " : "") + Array.prototype.slice.call(arguments, 2, 2 + n).join(", "));
+          },
+          toUpperCaseFromIndex(index, codePointIndex) {
+            const code = String.fromCodePoint(index).toUpperCase().codePointAt(codePointIndex|0);
+            return code !== undefined ? code : -1;
+          },
+          toLowerCaseFromIndex(index, codePointIndex) {
+            const code = String.fromCodePoint(index).toLowerCase().codePointAt(codePointIndex|0);
+            return code !== undefined ? code : -1;
```
Comment (dcodeIO) [inline tests/compiler.js]: There can be a `.js` file next to the respective test doing this on a per-test basis, so we don't have to add all of this to the main suite. See for example the `std/mod` test.

Comment (MaxGraey) [reply tests/compiler.js]: Good to know! Fixed!

File: std/assembly/string.ts [hunk 2]
```diff
@@ -484,6 +486,114 @@ import { idof } from "./builtins";
...
+    var len = <usize>this.length;
+    if (!len) return this;
+    var codes = __alloc(len * 2 * 2, idof<String>());
+    var j: usize = 0;
+    for (let i: usize = 0; i < len; ++i, ++j) {
+      let c = <u32>load<u16>(changetype<usize>(this) + (i << 1));
+      if (isAscii(c)) {
+        store<u16>(codes + (j << 1), toLower8(c));
+      } else {
+        if ((c - 0xD7FF < 0xDC00 - 0xD7FF) && i < len - 1) {
```
Comment (dcodeIO) [inline std/assembly/string.ts]: Can you add some comments here before the respective ifs, like which special cases are covered where? I don't for example understand the monkey patch.

File: std/assembly/util/casemap.ts [hunk 3]
```diff
@@ -0,0 +1,496 @@
...
+    let mid = (min + max) >>> 3 << 2;
+    let cmp = load<u16>(ptr + (mid << alignof<u16>())) - key;
+    if (cmp == 0) return mid; // found
+    else if (cmp >>> 31) min = mid + 4; // < 0
+    else max = mid - 4; // > 0
+  }
+  return -1; // not found
+}
+
+// See: musl/tree/src/ctype/towctrans.c
```
Comment (dcodeIO) [inline std/assembly/util/casemap.ts]: Did look at the musl file, but don't immediately see any similarities. Is this correct?

File: std/assembly/string.ts (hunk 2 above)
Comment (MaxGraey) [reply std/assembly/string.ts]: Ok, it;'s not monkey patch, just check and read second part of surrogate pair

File: std/assembly/util/casemap.ts (hunk 3 above)
Comment (MaxGraey) [reply std/assembly/util/casemap.ts]: Yeah, fixed

File: std/assembly/string.ts (hunk 2 above)
Comment (MaxGraey) [reply std/assembly/string.ts]: Fixed

File: std/assembly/string.ts [hunk 4]
```diff
@@ -484,6 +486,118 @@ import { idof } from "./builtins";
     // releases result
   }
 
+  toLowerCase(): String {
+    var len = <usize>this.length;
+    if (!len) return this;
+    var codes = __alloc(len * 2 * 2, idof<String>());
+    var j: usize = 0;
+    for (let i: usize = 0; i < len; ++i, ++j) {
```
Comment (dcodeIO) [inline std/assembly/string.ts]: Looks like removing the incrementor `++i, ++j` here and doing this in the respective cases below can save us one operation.

File: std/assembly/string.ts [hunk 5]
```diff
@@ -484,6 +486,118 @@ import { idof } from "./builtins";
...
+        if ((c - 0xD7FF < 0xDC00 - 0xD7FF) && i < len - 1) {
+          let c1 = <u32>load<u16>(changetype<usize>(this) + (i << 1), 2);
+          if (c1 - 0xDBFF < 0xE000 - 0xDBFF) {
+            let c0 = c;
+            c = (((c & 0x03FF) << 10) | (c1 & 0x03FF)) + 0x10000;
+            ++i;
+            if (c >= 0x20000) {
+              store<u32>(codes + (j << 1), c0 | (c1 << 16));
+              ++j;
+              continue;
```
Comment (dcodeIO) [inline std/assembly/string.ts]: Here for example it increments `i`, then `j` and then increments these again on `continue` using the incrementor.

File: std/assembly/string.ts [hunk 6]
```diff
@@ -484,6 +486,118 @@ import { idof } from "./builtins";
...
+            }
+          }
+        }
+        if (c == 0x0130) {
```
Comment (dcodeIO) [inline std/assembly/string.ts]: Can you explain what this checks for?

File: std/assembly/string.ts [hunk 7]
```diff
@@ -484,6 +486,118 @@ import { idof } from "./builtins";
...
+          store<u32>(codes + (j << 1), (0x0307 << 16) | 0x0069);
+          ++j;
+        } else if (c - 0x24B6 <= 0x24CF - 0x24B6) {
```
Comment (dcodeIO) [inline std/assembly/string.ts]: And this

File: std/assembly/string.ts [hunk 8]
```diff
@@ -484,6 +486,118 @@ import { idof } from "./builtins";
...
+            c = (((c & 0x03FF) << 10) | (c1 & 0x03FF)) + 0x10000;
+            ++i;
+            if (c >= 0x20000) {
+              store<u32>(codes + (j << 1), c0 | (c1 << 16));
+              ++j;
+              continue;
+            }
+          }
+        }
+        if (c - 0x24D0 <= 0x24E9 - 0x24D0) {
```
Comment (dcodeIO) [inline std/assembly/string.ts]: And this

File: std/assembly/string.ts (hunk 4 above)
Comment (MaxGraey) [reply std/assembly/string.ts]: Yes, but this significantly increase code size.

Comment (dcodeIO) [reply std/assembly/string.ts]: Hmm, would also help in spotting potential OOB accesses, which is what I was initially looking for. Wdyt, would making it a bit faster and clearer at the same time justify changing it?

Comment (MaxGraey) [reply std/assembly/string.ts]: In first variant I used exactly suggested by you version, but I rewrote to current due to better code size. In other case all other branches need add increments for `i` / `j`

Comment (dcodeIO) [reply std/assembly/string.ts]: Wondering how much of a difference this is. Especially considering that we are including relatively large tables anyway if these functions are used, this appears to be negligible to me. Also appears that an optimizer, at some point, might optimize to what we'd write manually today.

File: std/assembly/string.ts (hunk 5 above)
Comment (MaxGraey) [reply std/assembly/string.ts]: https://github.com/MaxGraey/assemblyscript/compare/case-mapping...MaxGraey:refactored-casemap?expand=1

File: std/assembly/string.ts (hunk 4 above)
Comment (MaxGraey) [reply std/assembly/string.ts]: https://github.com/MaxGraey/assemblyscript/compare/case-mapping...MaxGraey:refactored-casemap?expand=1

Comment (MaxGraey) [reply std/assembly/string.ts]: All right I could you that version. Mo big deal, but I'm not sure this got some performance benefits

Comment (MaxGraey) [reply std/assembly/string.ts]: Current version: https://webassembly.studio/?f=ul2gpz0nln
```
AS toLowerCase: 36.135009765625ms
AS toUpperCase: 34.879150390625ms
JS toLowerCase: 69.303955078125ms
JS toUpperCase: 65.5810546875ms
```

Suggested version: https://webassembly.studio/?f=mzyue5tfr9o
```
AS toLowerCase: 35.564208984375ms
AS toUpperCase: 36.23779296875ms
JS toLowerCase: 79.428955078125ms
JS toUpperCase: 84.647216796875ms
```

Comment (MaxGraey) [reply std/assembly/string.ts]: Btw on Firefox 71 suggested version 2.5x slower than current. So I prefer leave everithing as is

**Comment by dcodeIO** [comment]
Nicely done, thanks!

**Pull request merged**